README.rst
================
 Autoscale Test
================

Simple app and userdata to learn autoscaling.

Deploy a Pyramid app which gets load, perhaps by intense
calculations. Since we're planning on an API server, let's do this via
URL request returning JSON response. A factorial calculator might
do. Return metadata bout the server so we can tell who's answering the
request. Example::

  GET /factorial/5

  {
    "metadata": {"server" : "10.1.2.3"},
    "query": "/factorial/5",
    "answer": 120
  }

User Data contains entire install instructions
==============================================

For the first attempt, use a bare Ubuntu AMI and have the User Data do
all the work. In a real enviornment, this work would delay bringing up
the service quickly. But it means the server instance can be generic,
and all code updates will be reflected at instance launch time.

We supply to the instances (via AS Launch Config) userdata which
updates the OS, installs the prerequisites (e.g., pyramid), and pulls
the code from GitHub and integrates it with Pyramid. There must be
some system startup script too.


User Data contains almost nothing
=================================

For the second attempt, create an Ubuntu instance and manually update
the OS, install the prerequisites and code on the system. Then burn an
AMI from this.  Configure the Launch Config to use this AMI.  Is there
anything the User Data needs to do?


How this hierarchy was created, run
==============================

Create virtualenv.
bin/pcreate -s starter autoscaletest
bin/python setup.py develop
[hack autoscaletest to contain your app]
bin/pserve autoscaletest/development.ini 



CHANGES.rst
=========
 Changes
=========

2014-07-31
==========

Got app working.

Beating on it with 'ab' causes it to hang in AWS, and also my ssh
connection ot it. It this AWS blocking a perceived attack?



app.py
import platform
import logging

from wsgiref.simple_server import make_server
from pyramid.config import Configurator
from pyramid.response import Response

def _metadata():
    return {'metadata': {'node': platform.node()}}

def metadata(request):
    return _metadata()

def _factorial(number):
    # TODO: sanity check number < 1 and raise Exception
    # Recursive solution exceeded max depth between 500 and 1000, iterate.
    result = number
    for n in range (2, number):
        result *= n
    return result

def factorial(request):
    # TODO: Sanity check number is int
    number = int(request.matchdict['number'])
    md = _metadata()
    md.update({'number': number, 'factorial': _factorial(number)})
    return md

def _squareroot(number):
    # Newton's method
    DELTA = 1e-16
    steps = 0
    prev = 0                    # anything but z
    z = 1.0
    while abs(prev - z) > DELTA:
        prev = z
        steps += 1
        z -= (z * z - number) / (2 * z)
    return z, steps

def squareroot(request):
    number = int(request.matchdict['number'])
    sr, steps = _squareroot(number)
    md = _metadata()
    md.update({'number': number, 'squareroot': sr, 'steps': steps})
    return md

def home(request):
    return {'methods': ['/', '/metadata/{n}', '/squareroot/{n}', '/hello/{name}']}

def hello_world(request):
    md = _metadata()
    md.update({'msg': 'Hello %(name)s!' % request.matchdict})
    return md

if __name__ == '__main__':
    config = Configurator()

    config.add_route('home',       '/')
    config.add_route('metadata',   '/metadata')
    config.add_route('factorial',  '/factorial/{number}')
    config.add_route('squareroot', '/squareroot/{number}')
    config.add_route('hello',      '/hello/{name}')

    config.add_view(home,        route_name='home',       renderer='json')
    config.add_view(metadata,    route_name='metadata',   renderer='json')
    config.add_view(factorial,   route_name='factorial',  renderer='json')
    config.add_view(squareroot,  route_name='squareroot', renderer='json')
    config.add_view(hello_world, route_name='hello')

    app = config.make_wsgi_app()
    server = make_server('0.0.0.0', 8080, app)
    server.serve_forever()

# invoke as .../bin/python app.py





aws-ec2-userdata.sh
#!/bin/bash -x
# Runs as 'root' in '/'

# See AWS EC2 instance /var/log/cloud-init[-output].log
echo "######################################################################"
echo `date` RUNNING USER DATA AS USER `whoami` IN DIR `pwd`

# Ubuntu 14.04 has python-2.7
apt-get update -y
apt-get install -y git python-virtualenv

# TODO: use a tag
# https avoids git: ssh asking to trust the key, hanging AWS
echo "## Clone the repo and change into it..."
cd ~ubuntu
git clone https://github.com/shentonfreude/autoscale-pyramid
chown -R ubuntu:ubuntu autoscale-pyramid 
cd autoscale-pyramid

# Build and run with scripts in the code repo for easy changes
# since AWS doesn't let us edit the User Data. Do this as ubuntu user.
echo "## Invoke the build and run scripts..."
su ubuntu -c "cd ~ubuntu/autoscale-pyramid && ./build.sh"
# Best to run server in background
su ubuntu -c "cd ~ubuntu/autoscale-pyramid && ./run.sh"

echo `date` FINISHED USER DATA AND BUILDING / RUNNING
echo "######################################################################"


autoscaletest/__init__.py
import platform
import logging

from wsgiref.simple_server import make_server
from pyramid.config import Configurator
from pyramid.response import Response

def _metadata():
    return {'metadata': {'node': platform.node()}}

def metadata(request):
    return _metadata()

def _factorial(number):
    # TODO: sanity check number < 1 and raise Exception
    # Recursive solution exceeded max depth between 500 and 1000, iterate.
    result = number
    for n in range (2, number):
        result *= n
    return result

def factorial(request):
    # TODO: Sanity check number is int
    number = int(request.matchdict['number'])
    md = _metadata()
    md.update({'number': number, 'factorial': _factorial(number)})
    return md

def _squareroot(number):
    # Newton's method
    DELTA = 1e-9
    steps = 0
    prev = 0                    # anything but z
    z = 1.0
    while abs(prev - z) > DELTA:
        prev = z
        steps += 1
        z -= (z * z - number) / (2 * z)
    return z, steps

def squareroot(request):
    number = int(request.matchdict['number'])
    sr, steps = _squareroot(number)
    md = _metadata()
    md.update({'number': number, 'squareroot': sr, 'steps': steps})
    return md

def home(request):
    return {'methods': ['/', '/metadata/{n}', '/squareroot/{n}', '/hello/{name}']}

def hello_world(request):
    md = _metadata()
    md.update({'msg': 'Hello %(name)s!' % request.matchdict})
    return md

def main(global_config, **settings):
    config = Configurator()

    config.add_route('home',       '/')
    config.add_route('metadata',   '/metadata')
    config.add_route('factorial',  '/factorial/{number}')
    config.add_route('squareroot', '/squareroot/{number}')
    config.add_route('hello',      '/hello/{name}')

    config.add_view(home,        route_name='home',       renderer='json')
    config.add_view(metadata,    route_name='metadata',   renderer='json')
    config.add_view(factorial,   route_name='factorial',  renderer='json')
    config.add_view(squareroot,  route_name='squareroot', renderer='json')
    config.add_view(hello_world, route_name='hello')

    app = config.make_wsgi_app()
    return app


# Invoke as $virtualenv/bin/python autoscaletest/__init__.py

if __name__ == '__main__':
    app = main(None)
    server = make_server('0.0.0.0', 8080, app)
    server.serve_forever()

